### PR TITLE
Dropdown Menu layout respects menu items intrinsic sizes

### LINKED ISSUE #40773: DropdownMenuItems are fixed height
Internal: b/140830819

The height of dropdown menu items is fixed at 48. This is big enough to show one or two lines of text using the default font or one line of text with the default font and the maximum text scale factor on Android devices.

Menu items whose intrinsic heights are > 48 do not fit. This example demos the limitation:
https://gist.github.com/HansMuller/7df11ac474609a759b616d8258443f29

### PR DESCRIPTION
**Pull request by HansMuller**
Fixes https://github.com/flutter/flutter/issues/40773.

Previously, [dropdown menu items](https://api.flutter.dev/flutter/material/DropdownMenuItem-class.html) had a fixed height of 48 ([kMinInteractiveDimension](https://api.flutter.dev/flutter/material/kMinInteractiveDimension-constant.html)). Dropdown menu items now have a minimum size of 48  and no maximum size. When the menu appears, each item's vertical height will match its intrinsic height.

This [test case demos dropdown menus](https://gist.github.com/HansMuller/5e2277698678eebafe1956558b45277e) with variable height items. It also uses the new [selectedItemBuilder](https://github.com/flutter/flutter/pull/40461) feature to render compact versions of the selected items within the dropdown button.

dropdowns

| Previous   | Current    |         
| ------------- | ------------- |
|  dropdown1_open_fail |  dropdown1_open |
|  dropdown2_open_fail |  dropdown2_open |

### CODE REVIEW

**Review by shihaohong — COMMENTED** [review]
The changes look good! We should add docs to explain how to properly use this with `selectedItemBuilder` to get the desired effect multi-line effect, which could be a separate PR.

There's also an issue with the dropdown button itself not sizing properly with taller items, as you can see with this [gist](https://gist.github.com/shihaohong/378a75548e57e18ec3db49fec9c02033). 

Here's a screenshot with the problem:
Screen Shot 2019-09-25 at 5 25 41 PM

File: packages/flutter/lib/src/material/dropdown.dart [hunk 1]
```diff
@@ -457,6 +479,44 @@ class _DropdownRoutePage<T> extends StatelessWidget {
   }
 }
 
+// This widget only exists to enable _DropdownRoute to lookup the sizes of
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/dropdown.dart]: Sounds good, I updated the comment.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 2]
```diff
@@ -457,6 +479,44 @@ class _DropdownRoutePage<T> extends StatelessWidget {
...
+// of the dropdown button.
+class _MenuItem<T> extends SingleChildRenderObjectWidget {
+ const _MenuItem({
+    Key key,
+    @required this.onLayout,
+    @required this.item,
+  }) : assert(onLayout != null), assert(item != null), super(key: key, child: item);
+
+  final ValueChanged<Size> onLayout;
+  final DropdownMenuItem<T> item;
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/dropdown.dart]: It's used by _DropdownMenuState to look up the menu item's value.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 3]
```diff
@@ -898,12 +970,12 @@ class _DropdownButtonState<T> extends State<DropdownButton<T>> with WidgetsBindi
       items = widget.selectedItemBuilder == null
         ? List<Widget>.from(widget.items)
         : widget.selectedItemBuilder(context).map((Widget item) {
-            return Container(
-              height: _kMenuItemHeight,
-              alignment: AlignmentDirectional.centerStart,
-              child: item,
-            );
-          }).toList();
+          return Container(
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/dropdown.dart]: Good point, that wasn't indented properly. Fixed.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 4]
```diff
@@ -5,6 +5,7 @@
 import 'dart:math' as math;
 
 import 'package:flutter/widgets.dart';
+import 'package:flutter/rendering.dart';
```
Comment (gspencergoog) [inline packages/flutter/lib/src/material/dropdown.dart]: nit: "r" comes before "w", no?

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/lib/src/material/dropdown.dart (hunk 4 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/dropdown.dart]: Good point; reordered.

**Comment by HansMuller** [comment]
@shihaohong per https://github.com/flutter/flutter/pull/41120#pullrequestreview-292912693
This is a separate problem with dropdown buttons: currently the underline is drawn on top of the selected item, 8 pixels from the bottom of the selected item. It's not noticeable with the default menu item size. I'm going to write a separate PR to fix the underline.

**Review by shihaohong — APPROVED** [review]
LGTM

**Pull request merged**